File: src/contests/top150/n22_generate_parentheses/solution.go
package main

func generateParenthesis(n int) []string {
	var result []string
	backtracking(0, 0, n, "", &result)
	return result
}

func backtracking(open, close, n int, t string, result *[]string) {
	if open+close == 2*n {
		*result = append(*result, t)
	}

	if open < n {
		backtracking(open+1, close, n, t+"(", result)
	}
	if close < open {
		backtracking(open, close+1, n, t+")", result)
	}
}

func main() {
	result := generateParenthesis(1)
	for _, s := range result {
		println(s)
	}
}

File: src/contests/top150/n22_generate_parentheses/Solution.java
package contests.top150.n22_generate_parentheses;

import java.util.ArrayList;
import java.util.List;

class Solution {

    private List<String> result = new ArrayList();

    public List<String> generateParenthesis(int n) {
        backtracking(0, 0, n, "");
        return result;
    }

    private void backtracking(int open, int close, int n, String t) {
        if (open + close == 2 * n) {
            result.add(t);
        }

        if (open < n) { // (
            backtracking(open + 1, close, n, t + "(");
        }
        if (close < open) { // )
            backtracking(open, close + 1, n, t + ")");
        }
    }

    public static void main(String[] args) {
        var s = new Solution();
        s.generateParenthesis(3).forEach(System.out::println);
    }

}